### Reduction operation: the sum of the numbers in the range [0, a)

In [8]:
# IMPLEMENTACIÓN ORIGINAL

import time

def reduc_operation(a):
    """Compute the sum of the numbers in the range [0, a)."""
    x = 0
    for i in range(a):
        x += i
    return x

# Secuencial

value = 1000000

initialTime = time.time()
suma = reduc_operation(value)
finalTime = time.time()

print("Time taken by reduction operation:", (finalTime - initialTime), "seconds")

# Utilizando las operaciones mágicas de ipython
%timeit -r 2 reduc_operation(value)

print(f"\n \t Computing the sum of numbers in the range [0, value): {suma}\n")

Time taken by reduction operation: 0.03412318229675293 seconds
48 ms ± 4.43 μs per loop (mean ± std. dev. of 2 runs, 10 loops each)

 	 Computing the sum of numbers in the range [0, value): 499999500000



In [9]:
# 2. Usando listas de Python

# a) Crear lista con range
lista = list(range(value))
print(f"Lista creada con {len(lista)} elementos")

# b) Sumar con bucle for
print("\n--- Suma usando bucle for ---")
start_time = time.time()
suma_for = 0
for num in lista:
    suma_for += num
end_time = time.time()
print(f"Tiempo con bucle for: {end_time - start_time:.4f} segundos")
print(f"Resultado: {suma_for}")

# c) Sumar con función sum()
print("\n--- Suma usando función sum() ---")
start_time = time.time()
suma_func = sum(lista)
end_time = time.time()
print(f"Tiempo con función sum(): {end_time - start_time:.4f} segundos")
print(f"Resultado: {suma_func}")

# Comparar con timeit
print("\n--- Comparación con %timeit ---")
print("Bucle for:")
%timeit -r 2 -n 1 suma_for = 0; [suma_for := suma_for + num for num in lista]

print("\nFunción sum():")
%timeit -r 2 -n 1 sum(lista)

Lista creada con 1000000 elementos

--- Suma usando bucle for ---
Tiempo con bucle for: 0.0572 segundos
Resultado: 499999500000

--- Suma usando función sum() ---
Tiempo con función sum(): 0.0059 segundos
Resultado: 499999500000

--- Comparación con %timeit ---
Bucle for:
44.1 ms ± 921 μs per loop (mean ± std. dev. of 2 runs, 1 loop each)

Función sum():
5.86 ms ± 2.06 μs per loop (mean ± std. dev. of 2 runs, 1 loop each)


In [10]:
# 3. Usando arrays de NumPy

# %%
import numpy as np

# Convertir lista a array numpy
print("--- Usando NumPy ---")
array_np = np.array(lista)
print(f"Array numpy creado con {array_np.size} elementos")

# a) Sumar con bucle for
print("\n--- Suma con bucle for (array numpy) ---")
start_time = time.time()
suma_np_for = 0
for num in array_np:
    suma_np_for += num
end_time = time.time()
print(f"Tiempo con bucle for: {end_time - start_time:.4f} segundos")
print(f"Resultado: {suma_np_for}")

# b) Sumar con numpy.sum()
print("\n--- Suma con numpy.sum() ---")
start_time = time.time()
suma_np_func = np.sum(array_np)
end_time = time.time()
print(f"Tiempo con numpy.sum(): {end_time - start_time:.4f} segundos")
print(f"Resultado: {suma_np_func}")

# Comparar con timeit
print("\n--- Comparación con %timeit ---")
print("Bucle for en array numpy:")
%timeit -r 2 -n 1 suma_np_for = 0; [suma_np_for := suma_np_for + num for num in array_np]

print("\nnumpy.sum():")
%timeit -r 2 -n 1 np.sum(array_np)

--- Usando NumPy ---
Array numpy creado con 1000000 elementos

--- Suma con bucle for (array numpy) ---
Tiempo con bucle for: 0.0959 segundos
Resultado: 499999500000

--- Suma con numpy.sum() ---
Tiempo con numpy.sum(): 0.0005 segundos
Resultado: 499999500000

--- Comparación con %timeit ---
Bucle for en array numpy:
64.5 ms ± 1.02 ms per loop (mean ± std. dev. of 2 runs, 1 loop each)

numpy.sum():
457 μs ± 30.9 μs per loop (mean ± std. dev. of 2 runs, 1 loop each)


# CONCLUSIONES

## 1. Implementación Original
- Tiempo: 38.3 ms (34.3 ms ± 264 μs con timeit)
- Esta es la implementación más básica, se trata de un simple bucle for sobre range(). Aunque es sencilla, es relativamente eficiente porque range() genera números sobre la marcha sin crear una estructura de datos completa en memoria. Sin embargo, sigue siendo un bucle en Python puro, que es interpretado y por tanto más lento que código compilado.

## 2. Listas de Python
- Tiempo del bucle for sobre lista: 67.7 ms (42.8 ms ± 1.02 ms con timeit)
- Tiempo de la función sum(): 6.1 ms (6.04 ms ± 5.94 μs con timeit)
  
- El bucle es más lento que la implementación original (67.7 ms vs 38.3 ms). Esto se debe al tiempo que tarda en crear la lista completa en memoria (list(range(10⁶))) y en acceder a los elementos de la lista (más lento que iterar sobre range()).
- Por otro lado, sum() es mucho más rápida que el bucle (6.1 ms vs 67.7 ms) porque está implementada en C (no en Python), evita la sobrecarga del intérprete Python en cada iteración y está optimizada internamente para operaciones de reducción.

## 3. Arrays de NumPy
- Tiempo del bucle for sobre array: 116.5 ms (63.9 ms ± 872 μs con timeit)
- Tiempo de la función numpy.sum(): 0.5 ms (586 μs ± 9.13 μs con timeit)
- El bucle sobre array NumPy es 3 veces más lento que la implementación original. Esto ocurre porque los arrays NumPy tienen más sobrecarga para acceso individual, Python debe "desenvolver" cada elemento del array (boxing/unboxing) y se pierden las optimizaciones vectorizadas cuando se usa bucle Python.
- Finalmente, numpy.sum() es el MÁS RÁPIDO, 76 veces más que la implementación original (38.3 ms → 0.5 ms) y 12 veces más rápido que sum() de listas (6.1 ms → 0.5 ms). Esto se debe a que la operación vectorizada se ejecuta en C/Fortran, al uso de SIMD (instrucciones AVX/SSE) del procesador, la optimización de memoria contigua (arrays frente a listas) y la paralelización implícita en algunas operaciones.

In [14]:
# Para ejecución en terminal

import sys
import time
import numpy as np

# Función para ejecutar pruebas con un valor de N
def run_tests(N):
    print(f"\n=== EJECUCIÓN CON N = {N:,} ===")
    
    # 1. Original
    start = time.time()
    suma = 0
    for i in range(N):
        suma += i
    t_original = time.time() - start
    
    # 2. Lista + bucle for
    lista = list(range(N))
    
    start = time.time()
    suma_for = 0
    for num in lista:
        suma_for += num
    t_lista_for = time.time() - start
    
    # 3. Lista + sum()
    start = time.time()
    suma_sum = sum(lista)
    t_lista_sum = time.time() - start
    
    # 4. NumPy array + bucle for
    array_np = np.array(lista)
    
    start = time.time()
    suma_np_for = 0
    for num in array_np:
        suma_np_for += num
    t_np_for = time.time() - start
    
    # 5. NumPy + np.sum()
    start = time.time()
    suma_np = np.sum(array_np)
    t_np_sum = time.time() - start
    
    # Resultados
    print(f"\nTiempos de ejecución (segundos):")
    print(f"1. Original (for range):    {t_original:.6f}")
    print(f"2. Lista + for:             {t_lista_for:.6f}")
    print(f"3. Lista + sum():           {t_lista_sum:.6f}")
    print(f"4. NumPy array + for:       {t_np_for:.6f}")
    print(f"5. NumPy + np.sum():        {t_np_sum:.6f}")
    
    # Speedup
    speedup = t_lista_for / t_np_sum if t_np_sum > 0 else 0
    print(f"\nSpeedup (np.sum() vs lista+for): {speedup:.1f}x")
    
    # Verificación
    if suma == suma_for == suma_sum == suma_np_for == suma_np:
        print(f"\n✓ Resultado correcto: {suma:,}")
    else:
        print("\n✗ Error: resultados diferentes")
    
    return t_original, t_lista_for, t_lista_sum, t_np_for, t_np_sum

# Parámetro desde línea de comandos
if len(sys.argv) > 1:
    N = int(sys.argv[1])
    valores = [N]
    print(f"\nParámetro SLURM: N = {N:,}")
else:
    valores = [10**7, 10**8]
    print("\nValores por defecto: 10^7 y 10^8")

# Ejecutar pruebas
resultados = []
for valor in valores:
    tiempos = run_tests(valor)
    resultados.append((valor, *tiempos))

# Resumen final
print("\n" + "="*70)
print("RESUMEN FINAL")
print("="*70)
print(f"{'N':<12} {'Original':<10} {'Lista-for':<10} {'Lista-sum':<10} {'Array-for':<10} {'NumPy-sum':<10}")
print("-"*70)

for valor, t_orig, t_lfor, t_lsum, t_afor, t_nsum in resultados:
    print(f"{valor:<12,} {t_orig:<10.6f} {t_lfor:<10.6f} {t_lsum:<10.6f} {t_afor:<10.6f} {t_nsum:<10.6f}")

print("="*70)

ValueError: invalid literal for int() with base 10: '-f'

In [13]:
# Código para guardar el notebook
import json
import nbformat as nbf

print("Notebook completo creado.")
print("Guardar como: reduc-operation-alumno11.ipynb")

Notebook completo creado.
Guardar como: reduc-operation-alumno11.ipynb
